### PR TITLE
Add an I/O neurondb API

### PR DESCRIPTION
**Pull request by wizmer**
It introduces two classes:

## MorphInfo

A class containing all the information that can be found in the neurondb regarding a single morphology.

```python

info = MorphInfo(name='name', mtype='mtype', 'layer'='layer')
info.use_dendrite = False
info.axon_inputs = ['morph1', 'morph2']
```

## MorphDB

A class containing information regarding collections of morphologies.
The class has a single attribute `self.df` to expose this information as a
Pandas dataframe.

### Constructors:

```python
MorphDB()
MorphDB.from_neurondb()
MorphDB.from_folder()
```

### Adding new data

Data can be added through the the `+` and `+=` operators.

Example:

```python
all_morphs = MorphDB.from_neurondb('/gpfs/.../unravelled', label='unravelled')
all_morphs += MorphDB.from_neurondb('/gpfs/.../repaired', label='repaired')

more_morphs = MorphDB([MorphInfo(name='name', mtype='mtype', 'layer'='layer', label='extra-morphs'),
                       MorphInfo(name='name', mtype='mtype', 'layer'='layer', label='extra-morphs'),
                       MorphInfo(name='name', mtype='mtype', 'layer'='layer', label='extra-morphs'),
                       MorphInfo(name='name', mtype='mtype', 'layer'='layer', label='extra-morphs')])
morph2 = all_morphs + morph_info_list
```
### Writing neurondb to disk.

The data can be written to disk with both the XML and DAT format.

```python
MorphDB.from_neurondb('path1').write('neurondb.dat')
```

### Analysing morph-stats features

NeuroM morphometrics can be extracted with `self.features(config)` where
`config` is a [morph-stats
configuration](https://neurom.readthedocs.io/en/latest/morph_stats.html).

Labels can be used to compare morphometrics across different datasets.
```python
db = MorphDB.from_neurondb('path1', label='dataset-1')
db += MorphDB.from_neurondb('path2', label='dataset-2')
db += MorphDB.from_neurondb('path3', label='dataset-3')
db += MorphDB.from_neurondb('path4', label='dataset-4')

features = db.features(config)

for dataset, df in features.groupby(('neuron', 'label')):
    print(dataset, df)
```

### CODE REVIEW

File: morph_tool/morphdb.py
Comment (wizmer) [inline morph_tool/morphdb.py:188]: I'm tempted to change the name of this column to `region`. That would makes for sense for neurondb of hippocampus and thalamus.

File: morph_tool/morphdb.py
Comment (wizmer) [inline morph_tool/morphdb.py:206]: Maybe this flag should not appear at all. Or maybe be hidden by default ?

File: morph_tool/morphdb.py
Comment (mgeplf) [reply morph_tool/morphdb.py:188]: `region` usually applies to things like `MOs`, `FRM`, etc, so leaving it as layer is not optimal, but makes sense.  I've used `subregion` sometimes - but I don't think that's always clear.  I would poll the scientists, to see what they would like.

File: morph_tool/morphdb.py
Comment (mgeplf) [reply morph_tool/morphdb.py:206]: We'll need it to move away from `MorphologyTracker`, a better name would be helpful, to make its semantics clear.

How do you plan to make things hidden?

Comment (wizmer) [reply morph_tool/morphdb.py:206]: I'm balancing between:

- `show_deprecated` flag in the function signature. It would default to False
- Not exposing this flag at all. But that could be annoying the day we want to inspect something about it.
- Always expose it and let to the use the responsability to use the columns they need.

File: morph_tool/utils.py
Comment (arnaudon) [inline morph_tool/utils.py:36]: so this only reads .dat now? maybe update the docstring?

Comment (wizmer) [reply morph_tool/utils.py:36]: No, it also reads XML, see (new) line 38. Once again, GH diff formatting is weird.

File: morph_tool/morphdb.py
Comment (wizmer) [reply morph_tool/morphdb.py:188]: Elisabetta says in the case in the thalamus what appears under the `layer` tag are regions. I'm still waiting for Davide's answer.

> region usually applies to things like MOs, FRM

What are MOs, FRM and where can I see those terms ? Actual region from SSCX ?

File: morph_tool/utils.py
Comment (arnaudon) [reply morph_tool/utils.py:36]: ah indeed ;)

File: morph_tool/morphdb.py
Comment (arnaudon) [reply morph_tool/morphdb.py:188]: I would be more in favor or leaving the standard `layer` tag, as otherwise, most users will have to remember to change it  to `layer`, and people using `layer` as `region` are more aware of it that people using `layer` as `layer` (I think!)

File: morph_tool/morphdb.py
Comment (mgeplf) [reply morph_tool/morphdb.py:206]: I'm leaning towards
> Always expose it and let to the use the responsability to use the columns they need.

As long as the meaning is well documented.  It has the same maintenance problems as `show_deprecated`, since new/old properties will be added/deleted over time, so even if you set `show_deprecated=True`, you'd need to test what is actually returned.

File: morph_tool/morphdb.py
Comment (mgeplf) [reply morph_tool/morphdb.py:188]: The AIBS hierarchy looks like:
```
  "acronym": "FRP",                                                  
  "id": 184,                                                         
  "name": "Frontal pole, cerebral cortex",
  "children": [{
    "acronym": "FRP1",                                               
    "id": 68,                                                        
    "name": "Frontal pole, layer 1"                                  
   },                                                                
   {                                                                 
    "acronym": "FRP2",                                               
    "id": 20667,                                                     
    "name": "Frontal pole, layer 2"                                  
   },                                                                
  ....]
```
And, so far, we've considered 'FRP' to be a region, and its layers to be [1,2, 3..]: ie: leaf nodes are `layers` (`subregion`?), and regions are one level up.

The thalamus looks like:
```
  {                                                                 
   "acronym": "VP",                                                 
   "id": 709,                                                       
   "name": "Ventral posterior complex of the thalamus"              
   "children": [                                                    
    {                                                               
     "acronym": "VPL",                                              
     "id": 718,                                                     
     "name": "Ventral posterolateral nucleus of the thalamus"       
    },                                                              
    {                                                               
     "acronym": "VPLpc",                                            
     "id": 725,                                                     
     "name": "Ventral posterolateral nucleus of the thalamus, parvicellular part"
    },                                                              
    {                                                               
     "acronym": "VPM",                                              
     "id": 733,                                                     
     "name": "Ventral posteromedial nucleus of the thalamus"        
    },                                                              
    {                                                               
     "acronym": "VPMpc",                                            
     "id": 741,                                                     
     "name": "Ventral posteromedial nucleus of the thalamus, parvicellular part"
    }                                                               
   ],                                                               
  }, 
```
So, this *could* be considered to have region='VP', w/ `layers` / `subregion`s `VPL/VPM/...`

However, from what I remember from Elisabetta, she usually has `region` = VPL, and then would have no `subregion`.

This is to give a sense of the picture - maybe DKE (@annakristinkaufmann or @MFSY) can help w/ the terminology?  It would be good to use the standardization / create one.

File: morph_tool/morphdb.py
Comment (wizmer) [reply morph_tool/morphdb.py:206]: Yes and we can expect people to use only the columns they know/need/have a reason to use. The 'deprecated' tag in the docstring should deter people to use it without a good reason to do so.

File: morph_tool/morphdb.py
Comment (wizmer) [reply morph_tool/morphdb.py:188]: I guess (you probably know more about this Mike) that what is found in the current `layer` tag of the XML is the maximum amount of knowledge we have about the location of the morphology in the current state of things. That is, it could be anything between just a layer number and an actual region-subregion combo. In short I don't think the current neurondb.xml data we currently have are sufficient to correctly define what is expected in the `layer` tag.

So:
- we keep using `layer`. As Alexis said, non-sscx people are already used to this (for them) non-precise terminology. Maybe it's not the time to start changing the terminology. Moreover, this API has an intermediate lifetime: once Nexus is widespread neurondb.xml and the API will probably become obselete.
- we use `region` in the broader sense of things. After all, `layer 1`, `VP` or `VPL` can be considered 'brain regions'.

File: morph_tool/morphdb.py
Comment (mgeplf) [inline morph_tool/morphdb.py:48]: `item` is to generic a name, no?

dictionnary -> dictionary

File: morph_tool/morphdb.py
Comment (mgeplf) [inline morph_tool/morphdb.py:51]: It looks to me like it's easy to get things out of sync here: if someone assigns to `self.name`, the `self.item` dict isn't updated

File: morph_tool/morphdb.py
Comment (mgeplf) [inline morph_tool/morphdb.py:85]: why do we consider the hash to be the same when the subtypes/donors/etc differ?

File: morph_tool/morphdb.py
Comment (wizmer) [reply morph_tool/morphdb.py:51]: True. Well, I think we can actually discard this key. There is no real reason to keep the input data inside this class.

File: morph_tool/morphdb.py
Comment (wizmer) [reply morph_tool/morphdb.py:85]: This is based on `add_morphdb` and this is how it is decided if a morphology should be added to the morphdb or not.
```python
    def add_morph(self, name, mtype, layer, parent=None, dendrite=None, axon=None):
        '''add a morphology based on name, mtype and layer

        Add lineage information: if parent, then dendrite and axon are both set,
        or they are set independently

        If the name, mtype, layer combination already exists, then nothing is added

        Return:
            bool: True if morphology added, False otherwise
        '''
        morph_info = XMLMorphInfo(name=name, mtype=mtype, layer=layer,
                                  use_axon=None, use_dendrites=None,
                                  raw_xml=None)
        morph_key = MorphologyDB._morph_key(morph_info)
        if morph_key in self._known_morphologies:
            return False
```
However, after double checking, the mtype used is in fact the fullmtype so I will update the PR with that.

Comment (wizmer) [reply morph_tool/morphdb.py:85]: Maybe that's a bad design decision though. The logic involved in `MorphologyDB.add_morph` should not leak into designing the `MorphInfo` class.

**Review by arnaudon — DISMISSED** [review]

File: morph_tool/morphdb.py
Comment (arnaudon) [inline morph_tool/morphdb.py:55]: this is slow for large number of files, I had tried to bypass it, if the extension is known in advance

File: morph_tool/morphdb.py
Comment (arnaudon) [inline morph_tool/morphdb.py:83]: ok, so we will not use the terminology fullmtype at all? I'm good with it!

File: morph_tool/morphdb.py
Comment (arnaudon) [inline morph_tool/morphdb.py:234]: can't _from_xml be public?

File: morph_tool/morphdb.py
Comment (arnaudon) [inline morph_tool/morphdb.py:266]: we should allow to dump all the df, no?

File: morph_tool/morphdb.py
Comment (arnaudon) [inline morph_tool/morphdb.py:238]: I also added a line to put the abstolute morphology path as another column, would it make sense to have that? So i'll be easier later to load morphs, withoutt having to remember where the base_path is?

**Comment by arnaudon** [comment]
It looks really good, thanks @wizmer ! Now I'm wondering what is the best way to create a neurondb from scratch with this API? I will try to play with it in the context if MCAR, meaning I am given a bunch a morphollogies with some metadata, and I need to create a morph release, following curation, annotation, repair steps, and using all our codes. 
So I would need to create/edit the MorphDB along the way, to end up with final one at some point.
We also need to think of how to easily allow for other means of saving the DB, with nexus for example.

File: morph_tool/morphdb.py
Comment (adrien-berchet) [inline morph_tool/morphdb.py:142]: Shouldn't this function be a `@staticmethod` of `MorphDB`?

File: morph_tool/morphdb.py
Comment (wizmer) [reply morph_tool/morphdb.py:55]: Ah, yes. I forgot to put the `extension` argument back. I will do.

File: morph_tool/morphdb.py
Comment (wizmer) [reply morph_tool/morphdb.py:142]: Hum, yes I could do this

File: morph_tool/morphdb.py
Comment (wizmer) [reply morph_tool/morphdb.py:266]: I am not sure creating a new format is a good idea. If we need this, we'll do this in another review.

File: morph_tool/morphdb.py
Comment (arnaudon) [reply morph_tool/morphdb.py:55]: In fact, since we are here, we could add a method `.check_files_exist()` which runs this `path.exists()` and complain if one is missing. We could then use this method from times to times, to check nothing magically disappeard from gpfs :P

File: morph_tool/morphdb.py
Comment (wizmer) [reply morph_tool/morphdb.py:83]: Yes, I was thinking of using the `mtype` as the fullmtype. What do you think ?

**Comment by adrien-berchet** [comment]
Did you test it on folders with many morphologies? I am wondering how it scales for millions of morphologies.

File: morph_tool/morphdb.py
Comment (arnaudon) [reply morph_tool/morphdb.py:266]: Yes, we can always to `mydb.df.to_csv('all_data.csv')` for now

File: morph_tool/morphdb.py
Comment (wizmer) [reply morph_tool/morphdb.py:234]: Nope. `from_neurondb` is public so that's what should be used.

File: morph_tool/morphdb.py
Comment (arnaudon) [reply morph_tool/morphdb.py:83]: I like it. I've seen fullmtype in only a few places, such as in BluePyMM. I've never seen anyone needing the submtype of the mtyp_no_subtype at all, so I think it makes sense to stick to mtype and have  the  split as you  did (or make it optional even)

**Comment by arnaudon** [comment]
I was thinking about that, too @adrien-berchet , but I think we should use this API for small number of morphs, and use other tools for large numbers. This will be used for morphology release from biological reconstructions, and lab can reconstruct 9M morphs yet, so we are fine. For full synthesis of 9M cells, we won't really need this API, but rather the sonata format in h5, or other ones.

File: morph_tool/morphdb.py
Comment (wizmer) [reply morph_tool/morphdb.py:238]: This is a `Path` object so you should not need the absolute path I think.

Comment (arnaudon) [reply morph_tool/morphdb.py:238]: sorry, I was thinking of saving data to .csv, which I can carry around, so I need the absolute path saved in it.

**Comment by adrien-berchet** [comment]
Ah ok. If we are sure we will not need it for millions of cells it's cool.

**Comment by wizmer** [comment]
> Did you test it on folders with many morphologies? I am wondering how it scales for millions of morphologies.

No I did not but I don't think there should be any bottleneck. Reading from a DAT file is simply a call to `pd.read_csv` and the other expensive thing is to find the paths. But listing the content of the morphology folder can not really be avoided so I don't see much room for improvement.

Finally I did not add the extension back because it should no longer affect the performances. I have changed the way of looking upon them.

**Comment by adrien-berchet** [comment]
> No I did not but I don't think there should be any bottleneck. Reading from a DAT file is simply a call to `pd.read_csv` and the other expensive thing is to find the paths. But listing the content of the morphology folder can not really be avoided so I don't see much room for improvement.

I was more concerned about XML parsing. But maybe this is only used for biological data, so small data sets.
But we use it for the clones, right? These are quite big data sets. Though it should no more be used soon.

**Comment by wizmer** [comment]
In the morphlology repair workflow we are already using a similar parser, but it's true that we've never used it above a few 100 000 cells. We can always improve the perf in the future if we see that it's not enough.

**Comment by adrien-berchet** [comment]
:+1:

File: tests/test_morphdb.py
Comment (GianlucaFicarelli) [inline tests/test_morphdb.py:8]: Minor because we don't enforce linting on test files, but `numpy` and `eq_` are unused.
The file could be reformatted to be compliant to pep8.

File: morph_tool/morphdb.py
Comment (GianlucaFicarelli) [inline morph_tool/morphdb.py:296]: I'm not fully convinced that allowing to add together `MorphDB` and a list of `MorphInfo` is intuitive (I think that the two operands should usually belong to the same type, or to compatible types).
However I can change my mind. Maybe do you know other similar examples in this or other libraries?

File: morph_tool/morphdb.py
Comment (arnaudon) [inline morph_tool/morphdb.py:271]: this was a bit broken for me witht multi-index, as stats df did not have multi-index. I did that:
```
        df = self.df.copy().reset_index(drop=True)                                                  
        df.columns = pd.MultiIndex.from_product((["neuron"], df.columns.values))                    
        stats = extract_dataframe(df['neuron', 'path'], config, n_workers).drop(columns='name')     
        stats.columns = pd.MultiIndex.from_product((["stats"], stats.columns.values))               
        return df.join(stats, how='inner')  
```

Comment (arnaudon) [reply morph_tool/morphdb.py:271]: I am stupid, I was not using latest Neurom version, all good!

File: morph_tool/morphdb.py
Comment (wizmer) [reply morph_tool/morphdb.py]: Ah, yes, got it. Indeed, good point !

Comment (wizmer) [reply morph_tool/morphdb.py]: I think with the current way, they already are absolute paths.

Comment (arnaudon) [reply morph_tool/morphdb.py]: yes indeed, I just checked to be sure, all good!

File: morph_tool/morphdb.py
Comment (wizmer) [reply morph_tool/morphdb.py:105]: The problem with neurondb is that it is a massive mess and there is no specification so we are supposed to cover all the cases.
I've actually added a specification but most of the circuits were created back when there was no spec.

File: morph_tool/morphdb.py
Comment (wizmer) [reply morph_tool/morphdb.py:106]: No, empty strings are not needed.

File: morph_tool/morphdb.py
Comment (wizmer) [reply morph_tool/morphdb.py:228]: I've opted to keep the XML and DAT writers in this function but I've extracted the part that converts the data to a dictionary.

File: morph_tool/morphdb.py
Comment (wizmer) [reply morph_tool/morphdb.py:296]: I agree. I'll make the `MorphDB` ctor accept a sequence of `MorphInfo` object.
This way adding a list of MorphInfo and a MorphDB object will be:

```python
db2 = db1 + MorphDB(morph_info_seq)
```

Comment (GianlucaFicarelli) [reply morph_tool/morphdb.py:296]: If useful, you could add also an `update` method that can accept an iterable (it's not unusual to update a collection with an iterable).

File: morph_tool/morphdb.py
Comment (arnaudon) [inline morph_tool/morphdb.py:207]: maybe we should add here as a warning in docstring that we cannot have duplicate morphologies with different mtypes with this ctor. Or we can extend this with the option to give
 `mtypes: Dict[str, list[str]]`

File: morph_tool/morphdb.py
Comment (arnaudon) [inline morph_tool/morphdb.py:224]:
```suggestion
        output_path = Path(output_path)
        ext = output_path.suffix.lower()
```

File: morph_tool/morphdb.py
Comment (wizmer) [reply morph_tool/morphdb.py:207]: Yes good point. Or instead of `Dict[str, str]` we could use `Seq[Tuple[str, str]` so that passing a dict would still work (according that you use `.items()`) but also passing multiple lines corresponding to the same path.

Comment (arnaudon) [reply morph_tool/morphdb.py:207]: yes, that, too! I was trying to regenerate an neurondb.xml from a neurondb.dat... my god... this never ends...

File: morph_tool/morphdb.py
Comment (wizmer) [reply morph_tool/morphdb.py:296]: I think for the update `db += MorphDB(iterable)` is good enough, no ?

**Comment by arnaudon** [comment]
So with the `from_folder`, we can convert any other format of morphology db to this one, right? For example, if I just have a list of morphs in a .dat, I can create one, provided I give it a list of mtypes as well. 
This is good, it makes it very flexible at the user end.

**Comment by wizmer** [comment]
Yes, or you can also use the constructor
```python
db = MorphDB(MorphInfo(*data) for data in my_data_seq)
```

**Comment by arnaudon** [comment]
ah yes, sure! excellent! In my end, I think it's good to use, I've been trying it here and there. I'll be keen to merge it, so I can do some massive cleanup in our codes ;)

File: morph_tool/morphdb.py
Comment (GianlucaFicarelli) [inline morph_tool/morphdb.py:147]: I didn't notice before but the parameters should be represented with a repr (from the docs: "If at all possible, it should look like a valid Python expression that could be used to recreate an object with the same value").
```suggestion
        return (f'MorphInfo(name={self.name!r}, mtype={self.mtype!r}, layer={self.layer!r}, '
                f'label={self.label!r})')
```

File: morph_tool/morphdb.py
Comment (GianlucaFicarelli) [reply morph_tool/morphdb.py:168]: This has not been discussed or resolved.

File: morph_tool/morphdb.py
Comment (GianlucaFicarelli) [reply morph_tool/morphdb.py:190]: This has not been discussed or resolved.

File: morph_tool/morphdb.py
Comment (GianlucaFicarelli) [reply morph_tool/morphdb.py:294]: This has not been discussed or resolved.

File: morph_tool/morphdb.py
Comment (GianlucaFicarelli) [inline morph_tool/morphdb.py:246]: If I'm not wrong, a morphology folder can contain morphologies in different formats ('asc', 'h5', 'swc'), but we are keeping only one path (the last one returned by iter_morphology_files).
Is it ok? In that case, could you add a comment or an explanation in the docstring?

File: morph_tool/morphdb.py
Comment (GianlucaFicarelli) [reply morph_tool/morphdb.py:335]: This has not been discussed or resolved.

File: morph_tool/morphdb.py
Comment (GianlucaFicarelli) [inline morph_tool/morphdb.py:156]: I'm answering here to https://github.com/BlueBrain/morph-tool/pull/25#discussion_r565978296 because it's in outdated code.
Yes I think `db += MorphDB(iterable)` is good. The `update` method could be useful if some optimization is possible, but it seems to me that we need to concatenate the dataframes and create a new dataframe in either way.

Comment (GianlucaFicarelli) [inline morph_tool/morphdb.py:156]: The type should be `Iterable` instead of `Sequence` (it's called using a generator in `from_folder`).

Comment (wizmer) [reply morph_tool/morphdb.py:156]: Yes. Thank you

File: morph_tool/morphdb.py
Comment (GianlucaFicarelli) [inline morph_tool/morphdb.py:261]: I'm wondering if it could be more efficient to filter `files` based on the morphology names passed in `mtypes` (converted to a dict or set). It depends on the number of files, and the number of mtypes passed to the function.
However if there could be different cases I think it's fine as is, and we can improve it later if needed.

Comment (wizmer) [reply morph_tool/morphdb.py:261]: The general use case is that `iter_morphology_files` and `mtypes` should return the same files and should be the same length so that should not make a big difference I think.

**Review by GianlucaFicarelli — APPROVED** [review]

**Comment by wizmer** [comment]
Thank you !

**Comment by arnaudon** [comment]
🍾 !!!

**Comment by wizmer** [comment]
https://morph-tool.readthedocs.io/en/stable/morphdb.html

**Pull request merged**
